File: cloud-run-bot/advertising_compliance_service.js
/**
 * FırsatKolik - Ticari Reklam Yönetmeliği Uyum Servisi
 * 
 * 1 Ağustos Ticari Reklam Yönetmeliği gereğince, indirim/fırsat platformlarında
 * paylaşılan affiliate ve tanıtım içeriklerinin #tanıtım etiketi ile
 * etiketlenmesini sağlar. Varsa eski reklam/işbirliği etiketlerini #tanıtım'a dönüştürür.
 */

// Tanınan reklam/tanıtım/işbirliği kalıpları (hashtag'li, parantezli veya köşeli)
const ADVERTISING_TAG_CLEANUP_REGEX = /(?:#|\[|\()(?:reklam|reklamdır|reklamdir|tanıtım|tanitim|işbirliği|isbirligi|sponsorlu|ortaklık|ortaklik|affiliate)(?:\]|\))?/gi;
const ADVERTISING_REGEX = /(?:#|\b)(?:reklam|reklamdır|reklamdir|tanıtım|tanitim|işbirliği|isbirligi|sponsorlu|ortaklık|ortaklik|affiliate)\b/i;

/**
 * Verilen metinde reklam/tanıtım/işbirliği etiketinin olup olmadığını kontrol eder.
 * @param {string} text 
 * @returns {boolean}
 */
function hasAdvertisingDisclosure(text) {
  if (!text || typeof text !== 'string') return false;
  
  const trimmed = text.trim();
  if (!trimmed) return false;

  const lowerStandard = trimmed.toLowerCase();
  const lowerTurkish = trimmed.toLocaleLowerCase('tr-TR');

  return ADVERTISING_REGEX.test(lowerStandard) || ADVERTISING_REGEX.test(lowerTurkish);
}

/**
 * Açıklama metnini normalize eder:
 * - Varsa eski #reklam, #işbirliği, [REKLAM] vb. etiketleri temizler.
 * - Metnin sonuna standart tek bir '#tanıtım' etiketi ekler.
 * 
 * @param {string} text - Orijinal açıklama metni
 * @param {string} defaultTag - Eklenecek standart etiket (Varsayılan: '#tanıtım')
 * @returns {string}
 */
function ensureAdvertisingDisclosure(text, defaultTag = '#tanıtım') {
  let safeText = (text || '').trim();

  // Eğer metin tamamen boşsa varsayılan açıklama + etiket dön
  if (!safeText) {
    return `Fırsat Ürünü Detayları\n\n${defaultTag}`;
  }

  // Varsa eski reklam/tanıtım/işbirliği etiketlerini kaldır
  safeText = safeText.replace(ADVERTISING_TAG_CLEANUP_REGEX, '').trim();
  safeText = safeText.replace(/\n{3,}/g, '\n\n').trim();

  if (!safeText) {
    return `Fırsat Ürünü Detayları\n\n${defaultTag}`;
  }

  // Sonuna standart #tanıtım etiketini ekle
  return `${safeText}\n\n${defaultTag}`;
}

module.exports = {
  hasAdvertisingDisclosure,
  ensureAdvertisingDisclosure,
  ADVERTISING_REGEX,
  ADVERTISING_TAG_CLEANUP_REGEX,
};

File: lib/services/advertising_compliance_service.dart
/// FırsatKolik - Ticari Reklam Yönetmeliği Uyum Servisi
/// 
/// 1 Ağustos Ticari Reklam Yönetmeliği gereğince, indirim/fırsat platformlarında
/// paylaşılan affiliate ve bilgilendirme içeriklerinin #tanıtım etiketi ile
/// etiketlenmesini sağlar. Varsa eski #reklam/#işbirliği etiketlerini #tanıtım'a çevirir.
class AdvertisingComplianceService {
  // Temizlenecek eski reklam/tanıtım/işbirliği kalıpları
  static final RegExp _adCleanupRegex = RegExp(
    r'(?:#|\[|\()(?:reklam|reklamdır|reklamdir|tanıtım|tanitim|işbirliği|isbirligi|sponsorlu|ortaklık|ortaklik|affiliate)(?:\]|\))?',
    caseSensitive: false,
  );

  static final RegExp _adRegex = RegExp(
    r'(?:#|\b)(?:reklam|reklamdır|reklamdir|tanıtım|tanitim|işbirliği|isbirligi|sponsorlu|ortaklık|ortaklik|affiliate)\b',
    caseSensitive: false,
  );

  /// Verilen metinde reklam/tanıtım/işbirliği etiketinin olup olmadığını kontrol eder.
  static bool hasDisclosure(String? text) {
    if (text == null) return false;
    final trimmed = text.trim();
    if (trimmed.isEmpty) return false;

    return _adRegex.hasMatch(trimmed.toLowerCase()) || _adRegex.hasMatch(trimmed);
  }

  /// Açıklama metnini normalize eder:
  /// - Varsa eski #reklam, #işbirliği, [REKLAM] vb. etiketleri temizler.
  /// - Metnin sonuna standart tek bir '#tanıtım' etiketi ekler.
  static String ensureDisclosure(String? text, {String defaultTag = '#tanıtım'}) {
    if (text == null) return defaultTag;
    var safeText = text.trim();

    if (safeText.isEmpty) {
      return defaultTag;
    }

    // Var olan eski etiketleri temizle
    safeText = safeText.replaceAll(_adCleanupRegex, '').trim();
    safeText = safeText.replaceAll(RegExp(r'\n{3,}'), '\n\n').trim();

    if (safeText.isEmpty) {
      return defaultTag;
    }

    // Standart #tanıtım etiketini ekle
    return '$safeText\n\n$defaultTag';
  }
}
